sklearn_numba_dpex/kmeans/kernels.py

```diff
@@ -0,0 +1,620 @@
+                previous_center[feature_idx, sample_idx]
+                - center[feature_idx, sample_idx]
+            )
+            tmp += center_diff * center_diff
+
+        center_shift[sample_idx] = tmp
+
+    return center_shift[global_size, work_group_size]
+
+
+@lru_cache
+def make_half_l2_norm_dim0_kernel(n_samples, n_features, work_group_size, dtype):
+    global_size = math.ceil(n_samples / work_group_size) * work_group_size
+    zero = dtype(0.0)
+
+    @dpex.kernel
+    def half_l2_norm(X, result):
+        sample_idx = dpex.get_global_id(0)
+
+        if sample_idx >= n_samples:
+            return
+
+        l2_norm = zero
+
+        for feature_idx in range(n_features):
+            item = X[feature_idx, sample_idx]
+            l2_norm += item * item
+
+        result[sample_idx] = l2_norm / 2
+
+    return half_l2_norm[global_size, work_group_size]
+
+
+@lru_cache
+def make_sum_reduction_1dim_kernel(n_samples, work_group_size, device, dtype):
+    local_nb_iterations = math.floor(math.log2(work_group_size))
+    zero = dtype(0.0)
+
+    @dpex.kernel
+    def sum_reduction_kernel(v, w):
+        group_id = dpex.get_group_id(0)
+        local_work_id = dpex.get_local_id(0)
+        first_work_id = local_work_id == 0
+
+        n_samples = v.shape[0]
+
+        shm = dpex.local.array(work_group_size, dtype=dtype)
+
+        first_sample_idx = group_id * work_group_size * 2
+        augend_idx = first_sample_idx + local_work_id
+        addend_idx = first_sample_idx + work_group_size + local_work_id
+
+        if augend_idx >= n_samples:
+            shm[local_work_id] = zero
+        elif addend_idx >= n_samples:
+            shm[local_work_id] = v[augend_idx]
+        else:
+            shm[local_work_id] = v[augend_idx] + v[addend_idx]
+
+        dpex.barrier()
+        current_size = work_group_size
+        for i in range(local_nb_iterations):
+            current_size = current_size // 2
+            if local_work_id < current_size:
+                shm[local_work_id] += shm[local_work_id + current_size]
+
+            dpex.barrier()
+
+        if first_work_id:
+            w[group_id] = shm[0]
+
+    _steps_data = []
+    n_groups = n_samples
+    while n_groups > 1:
+        n_groups = math.ceil(n_groups / (2 * work_group_size))
+        global_size = n_groups * work_group_size
+        _steps_data.append(
+            (
+                sum_reduction_kernel[global_size, work_group_size],
+                dpctl.tensor.empty(n_groups, dtype=dtype, device=device),
+            )
+        )
+
+    def sum_reduction(v):
+        for sum_fn, w in _steps_data:
+            sum_fn(v, w)
+            v = w
+        return v
+
+    return sum_reduction
+
+
+@lru_cache
+def make_sum_reduction_2dim_kernel(n_samples, n_features, work_group_size, dtype):
+    global_size = math.ceil(n_samples / work_group_size) * work_group_size
+    zero = dtype(0.0)
+
+    @dpex.kernel
+    def sum_reduction_kernel(v, w):
```

Comment:
v and w should be renamed to more explicit names related to the actual use of this kernel in our LLoyd driver.

Reply:
The kernel function itself should be renamed to be more specific. Let's not pretend that those kernels are generic if we need to use them only once in our code.

Reply:
About those kernels, it's not that we will re-use them but I thought that describing a sum as a sum rather than a sum of two particular things is actually better for understanding ? Beside, the intent of those drivers is made clear (with specific naming) when used in the driver.

Reply:
ok for generic names but then drop `n_samples` and `n_features` and put the axis along which the reduction happen in the kernel name.

Reply:
do you have any suggestions for `n_samples`, `n_features` ? going for `size0`, `size1`, ... , `sizen`